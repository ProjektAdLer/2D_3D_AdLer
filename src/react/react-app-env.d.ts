/// <reference types="react-scripts" />

// Images
declare module '*.jpg';
declare module '*.png';
declare module '*.env';

// 3D types
declare module '*.glb';
declare module '*.stl';

declare module '*.jpeg';
declare module '*.gif';
declare module '*.mp4';
