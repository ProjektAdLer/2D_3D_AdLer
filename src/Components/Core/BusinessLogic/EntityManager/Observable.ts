export default abstract class Observable<T> {
  protected value: T;
  protected subscribers: ((value: T) => void)[] = [];

  public subscribe(fn: (value: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (value: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  protected notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  get Value(): T {
    return this.value;
  }
}
