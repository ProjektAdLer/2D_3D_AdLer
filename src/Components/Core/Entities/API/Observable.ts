export default class Observable<T> {
  constructor(data?: T) {
    if (data) {
      this.data = data;
    }
  }
  private data: T;
  private subscribers: Array<(data: T) => void> = [];

  public subscribe(fn: (data: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (data: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  public notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  public getData(): T {
    return this.data;
  }

  public setData(data: T): void {
    this.data = data;
    this.notify(data);
  }
}
