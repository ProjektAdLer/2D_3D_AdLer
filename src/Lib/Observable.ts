export default class Observable<T> {
  protected value: T;
  protected subscribers: ((value: T) => void)[] = [];

  constructor(defaultValue?: T, protected notifyOnSameValue: boolean = false) {
    if (defaultValue) this.value = defaultValue;
  }

  public subscribe(fn: (value: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (value: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  protected notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  get Value(): T {
    return this.value;
  }

  set Value(value: T) {
    if (this.notifyOnSameValue || value !== this.value) {
      this.value = value;
      this.notify(value);
    }
  }
}
