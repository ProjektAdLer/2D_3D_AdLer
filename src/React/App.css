body {
  margin: 0;
  height: 100vh;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

div{
  width: 100%;
  height: 100%;
}

section{
  display: inline;
  position: fixed;
  top: 0;
  left: 0;
  width: 120px;
  height: 100vh;
  background-color: #fff;
}

button{
  width: 100px;
  margin: 10px;
  z-index: 5;
  position: fixed;
  left: 0;
  top: 0;
}

canvas {
  width: calc(100% - 100px);
  height: 100%;
  margin-left: 100px;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

.modal {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.5);
  display: flex;
  align-items: center;
  justify-content: center;
}

.modal-content {
  width: 1000px;
  background-color: #fff;
}

.modal-header,
.modal-footer {
  padding: 10px;
  height: 60px;
}

.modal-title {
  margin: 0;
}

.modal-body {
  padding: 10px;
  border-top: 1px solid #eee;
  border-bottom: 1px solid #eee;
}
