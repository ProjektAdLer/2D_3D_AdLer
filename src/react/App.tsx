import React from 'react';
import './App.css';
import BabylonCanvas from './components/BabylonCanvas';

function App() {
	return (
		<React.Fragment>
			<BabylonCanvas style={{ width: '100%', position: 'absolute' }} />
		</React.Fragment>
	);
}

export default App;
