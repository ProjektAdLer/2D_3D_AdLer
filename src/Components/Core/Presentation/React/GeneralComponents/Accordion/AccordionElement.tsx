import { useEffect, useState } from "react";
import Observable from "src/Lib/Observable";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";
import tailwindMerge from "../../../Utils/TailwindMerge";

export interface AccordionProps {
  isOpen: Observable<boolean>;
  content: JSX.Element;
  header?: string;
  className?: string;
  buttonClassName?: string;
  headerClassName?: string;
  toolTip?: string;
}

export default function AccordionElement(props: AccordionProps) {
  const [isOpen, setIsOpen] = useObservable<boolean>(props.isOpen);
  const [arrowSymbol, setArrowSymbol] = useState<string>(
    props.isOpen.Value ? "\u25BC" : "\u25B6",
  );

  // content as useState prevents occassional pop-in of thumbnails ~sb
  const [content, setContent] = useState<JSX.Element | null>(
    props.isOpen.Value ? props.content : null,
  );
  const [cssEffect, setCssEffect] = useState<string>("");

  useEffect(() => {
    isOpen ? setArrowSymbol("\u25BC") : setArrowSymbol("\u25B6");
    isOpen ? setContent(props.content) : setContent(null);
    // prevents repeated opening-animation on already open accordion elements on category switch ~sb
    if (isOpen === props.isOpen.Value) {
      setCssEffect(isOpen ? "accordionOpening" : "accordionCollapsing");
    } else {
      setCssEffect("");
    }
  }, [isOpen, props]);

  return (
    <>
      <div
        className={tailwindMerge(
          "pb-2 border-b border-gray-500",
          props.className ?? "",
        )}
      >
        <button
          className={tailwindMerge(
            "content-start w-full text-left",
            props.buttonClassName ?? "",
          )}
          onClick={() => {
            setIsOpen(!isOpen);
          }}
          title={props.toolTip}
        >
          <h1
            className={tailwindMerge(
              "text-2xl portrait:text-lg font-bold",
              props.headerClassName ?? "",
            )}
          >
            {arrowSymbol + " " + props.header}
          </h1>
        </button>
      </div>
      <div>
        <div className={cssEffect}>{content}</div>
      </div>
    </>
  );
}
