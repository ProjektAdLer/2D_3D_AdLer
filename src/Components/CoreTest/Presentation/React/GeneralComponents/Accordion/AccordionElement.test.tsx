import { fireEvent, render, waitFor } from "@testing-library/react";
import AccordionElement from "../../../../../Core/Presentation/React/GeneralComponents/Accordion/AccordionElement";
import React from "react";
import Observable from "../../../../../../Lib/Observable";

describe("AccordionElement", () => {
  test("should render", () => {
    const container = render(
      <AccordionElement
        isOpen={new Observable<boolean>(true)}
        content={<></>}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("content renders when isopen is true", () => {
    const container = render(
      <AccordionElement
        isOpen={new Observable<boolean>(true)}
        content={<div>test</div>}
      />,
    );
    waitFor(() => {
      expect(container.queryAllByDisplayValue("test").length).toBe(1);
    });
  });

  test("content wont render when isopen is false", () => {
    const container = render(
      <AccordionElement
        isOpen={new Observable<boolean>(false)}
        content={<div>test</div>}
      />,
    );
    waitFor(() => {
      expect(container.queryAllByDisplayValue("test").length).toBe(0);
    });
  });

  test("clicking on button opens/closes content", () => {
    const container = render(
      <AccordionElement
        isOpen={new Observable<boolean>(false)}
        content={<div>test</div>}
      />,
    );

    container.queryByRole("button")?.click();

    waitFor(() => {
      expect(container.queryAllByDisplayValue("test").length).toBe(1);
    });
  });
});
