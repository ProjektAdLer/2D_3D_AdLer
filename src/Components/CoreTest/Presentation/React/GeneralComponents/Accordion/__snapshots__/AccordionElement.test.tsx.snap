// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionElement should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pb-2 border-b border-gray-500"
      >
        <button
          class="content-start w-full text-left"
        >
          <h1
            class="text-2xl portrait:text-lg font-bold"
          >
            ▼ undefined
          </h1>
        </button>
      </div>
      <div>
        <div
          class="accordionOpening"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pb-2 border-b border-gray-500"
    >
      <button
        class="content-start w-full text-left"
      >
        <h1
          class="text-2xl portrait:text-lg font-bold"
        >
          ▼ undefined
        </h1>
      </button>
    </div>
    <div>
      <div
        class="accordionOpening"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
